README.md
# sevenwonders

An app aimed to simplify and speed up the scoring process for the well known boardgame **[7 Wonders](https://boardgamegeek.com/boardgame/316377/7-wonders-second-edition)**. 

It also supports expansions : [Leaders](https://boardgamegeek.com/boardgameexpansion/316380/7-wonders-second-edition-leaders) | [Cities](https://boardgamegeek.com/boardgameexpansion/316378/7-wonders-second-edition-cities) | [Armada](https://boardgamegeek.com/boardgameexpansion/316382/7-wonders-second-edition-armada).


CHANGELOG.md
# Changelog

## [1.0.1] - 2021-03-15

### Changed

- Score sheet design.

### Fixed

- Guilde score when Armada is enabled.
- Maximum and minimum possible scores.

## [1.0.0] - 2021-03-14

### Added

- Score Sheet for 7 Wonders boardgame.
- Expansion Leaders.
- Expansion Cities.
- Expansion Armada.

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:sevenwonders/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/models/player.dart
import 'dart:math';

class Player {
  String? name;

  int warScore = 0;
  int wonderScore = 0;
  int civilianScore = 0;
  int commerceScore = 0;
  int guildeScore = 0;

  int money = 0;
  int debt = 0;

  // science
  int gears = 0;
  int compass = 0;
  int tablets = 0;
  int wilds = 0;

  int leaders = 0;
  int cities = 0;
  int armada = 0;

  Player(String name) {
    this.name = name;
  }

  int totalScore(bool leaders, cities, armada) {
    var total = this.warScore +
        this.wonderScore +
        this.civilianScore +
        this.commerceScore +
        (armada ? min(10, this.guildeScore) : this.guildeScore) +
        (leaders ? this.leaders : 0) +
        (cities ? this.cities : 0) +
        _moneyScore(this.money) +
        (cities ? this.debt : 0) +
        (armada ? this.armada : 0) +
        _scienceScore(this.wilds, this.compass, this.gears, this.tablets);
    return total as int;
  }

  static int _moneyScore(int m) {
    return m ~/ 3;
  }

  // _scienceScore calcultation according to wilds, compass, gears, and tablets.
  static int _scienceScore(int w, int c, int g, int t) {
    if (w <= 0) {
      return (g * g) + (c * c) + (t * t) + 7 * (min(min(c, g), t));
    }
    return max(
        max(_scienceScore(w - 1, c + 1, g, t),
            _scienceScore(w - 1, c, g + 1, t)),
        _scienceScore(w - 1, c, g, t + 1));
  }
}


lib/components/selector.dart
import 'package:flutter/material.dart';

import 'package:numberpicker/numberpicker.dart';

class NumberSelector extends StatelessWidget {
  const NumberSelector({
    Key? key,
    required this.value,
    this.min,
    this.max,
    required this.onChange,
  }) : super(key: key);

  final int value;

  final int? max;

  final int? min;

  final ValueChanged<int?>? onChange;

  @override
  Widget build(BuildContext context) {
    return NumberPicker(
      value: this.value,
      minValue: this.min ?? 0,
      maxValue: this.max ?? 999,
      textStyle: TextStyle(fontSize: 14, color: Colors.black.withOpacity(0.5)),
      selectedTextStyle: TextStyle(fontSize: 16),
      itemWidth: 50,
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(16),
        border: Border.all(color: Colors.black26),
      ),
      onChanged: (v) => onChange!(v),
      axis: Axis.horizontal,
    );
  }
}


lib/components/list_tile.dart
import 'package:flutter/material.dart';

import 'package:sevenwonders/components/selector.dart';

class ListTilePlayerWithNumberSelector extends StatelessWidget {
  const ListTilePlayerWithNumberSelector({
    Key? key,
    required this.playerName,
    required this.selector,
  })   : assert(playerName.length > 0),
        super(key: key);

  final String playerName;

  final NumberSelector selector;

  @override
  Widget build(BuildContext context) {
    return new Container(
      child: Row(
        children: [
          Expanded(
            child: ListTile(
              contentPadding:
                  EdgeInsets.symmetric(vertical: 5.0, horizontal: 16.0),
              title: new Text('${this.playerName}'),
            ),
          ),
          selector,
        ],
      ),
    );
  }
}


lib/screens/score/score.dart
import 'dart:collection';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'package:sevenwonders/models/player.dart';
import 'package:sevenwonders/components/selector.dart';
import 'package:sevenwonders/components/list_tile.dart';

class ScoreSheetPage extends StatefulWidget {
  final String title = "Score Sheet";

  @override
  ScoreSheetPageState createState() => ScoreSheetPageState();
}

class ScoreSheetPageState extends State<ScoreSheetPage> {
  List<Player> players = [];

  bool exLeaders = false;
  bool exCities = false;
  bool exArmada = false;

  @override
  Widget build(BuildContext context) {
    var sections = _tabsSections();

    return DefaultTabController(
      length: sections.length,
      child: Scaffold(
        drawer: _drawer(),
        appBar: AppBar(
          title: Text("Feuille des scores"),
          actions: <Widget>[
            _actionNewGame(context),
          ],
          bottom: TabBar(
            isScrollable: true,
            tabs: sections.keys.toList(),
          ),
        ),
        body: TabBarView(
          children: sections.values.toList(),
        ),
      ),
    );
  }

  LinkedHashMap<Widget, Widget> _tabsSections() {
    LinkedHashMap<Widget, Widget> sections = LinkedHashMap();
    sections.addAll({
      Tab(text: "JOUEURS"): _playersListView(),
      Tab(text: "MILITAIRE"): _listView(_playerWarView),
      Tab(text: "MONNAIE"): _listView(_playerMoneyView),
    });
    if (this.exCities == true) {
      sections[Tab(text: "DETTE")] = _listView(_playerDebtView);
    }
    sections.addAll({
      Tab(text: "MERVEILLE"): _listView(_playerWonderView),
      Tab(text: "CIVIL"): _listView(_playerCivilianView),
      Tab(text: "COMMERCE"): _listView(_playerCommerceView),
      Tab(text: "SCIENCE"): _listViewScience(_playerScienceView),
      Tab(text: "GUILDE"): _listView(_playerGuildeView),
    });
    if (this.exLeaders == true) {
      sections[Tab(text: "LEADERS")] = _listView(_playerLeadersView);
    }
    if (this.exCities == true) {
      sections[Tab(text: "CITIES")] = _listView(_playerCitiesView);
    }
    if (this.exArmada == true) {
      sections[Tab(text: "ARMADA")] = _listView(_playerArmadaView);
    }
    sections[Tab(text: "TOTAL")] = _listViewSorted(_playerTotalView);
    return sections;
  }

  Drawer _drawer() {
    return Drawer(
      child: new Container(
        padding: EdgeInsets.only(top: 50),
        child: new Column(
          children: <Widget>[
            ListTile(
              title: Text(
                'Extensions'.toUpperCase(),
                style: TextStyle(
                  fontSize: 16,
                ),
              ),
              onTap: null,
            ),
            Column(
              children: <Widget>[
                CheckboxListTile(
                  title: Text("Leaders"),
                  value: this.exLeaders,
                  onChanged: (val) {
                    setState(() {
                      exLeaders = val ?? false;
                    });
                  },
                ),
                CheckboxListTile(
                  title: Text("Cities"),
                  value: this.exCities,
                  onChanged: (val) {
                    setState(() {
                      exCities = val ?? false;
                    });
                  },
                ),
                CheckboxListTile(
                  title: Text("Armada"),
                  value: this.exArmada,
                  onChanged: (val) {
                    setState(() {
                      this.exArmada = val ?? false;
                    });
                  },
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }

  TextButton _actionNewGame(BuildContext context) {
    return TextButton(
      child: Icon(Icons.refresh, color: Colors.white),
      onPressed: () {
        showDialog(
          context: context,
          builder: (BuildContext context) {
            return new AlertDialog(
              title: new Text('Nouvelle partie'),
              content:
                  new Text('Effacer les scores et recommencer une partie ?'),
              actions: <Widget>[
                new TextButton(
                    child: new Text('NON'),
                    onPressed: () => Navigator.of(context).pop()),
                new TextButton(
                  child: Text("OUI"),
                  onPressed: () {
                    setState(() {
                      players = [];
                    });
                    Navigator.of(context).pop();
                  },
                ),
              ],
            );
          },
        );
      },
    );
  }

  Widget _playersListView() {
    return Scaffold(
      body: _listView(_playerView),
      floatingActionButton: new FloatingActionButton(
        backgroundColor: Colors.blue,
        onPressed: _addPlayerLayout,
        child: new Icon(Icons.add),
      ),
    );
  }

// _listView display a player list with a function for all player list tile
  Widget _listView(Widget Function(int, Player) playerView) {
    return new Card(
      margin: const EdgeInsets.only(top: 20.0, left: 10.0, right: 10.0),
      child: ListView.separated(
        separatorBuilder: (context, index) => Divider(
          color: Colors.black12,
        ),
        itemCount: players.length,
        itemBuilder: (context, index) {
          return playerView(index, players[index]);
        },
      ),
    );
  }

  // _listView display a player list with a function for all player list tile
  Widget _listViewScience(Widget Function(int, Player) playerView) {
    return new Card(
      margin: const EdgeInsets.only(top: 10.0, left: 10.0, right: 10.0),
      child: ListView.separated(
        separatorBuilder: (context, index) => Divider(
          color: Colors.black12,
        ),
        itemCount: players.length + 1,
        itemBuilder: (context, index) {
          if (index == 0) {
            return Container(
              child: Row(
                children: [
                  Spacer(flex: 2),
                  Expanded(child: Text("compas", textAlign: TextAlign.center)),
                  Expanded(child: Text("roue", textAlign: TextAlign.center)),
                  Expanded(child: Text("pierre", textAlign: TextAlign.center)),
                  Expanded(child: Text("bonus", textAlign: TextAlign.center)),
                ],
              ),
            );
          }
          return playerView(index - 1, players[index - 1]);
        },
      ),
    );
  }

  Widget _listViewSorted(Widget Function(int, Player) playerView) {
    List<Player> playersSorted = new List.from(players);
    playersSorted.sort((a, b) {
      var ascore = a.totalScore(this.exLeaders, this.exCities, this.exArmada);
      var bscore = b.totalScore(this.exLeaders, this.exLeaders, this.exArmada);
      if (ascore < bscore) {
        return 1;
      } else if (ascore > bscore) {
        return -1;
      } else {
        return -Comparable.compare(a.money, b.money);
      }
    });
    return new Card(
      margin: const EdgeInsets.only(top: 20.0, left: 10.0, right: 10.0),
      child: ListView.separated(
        separatorBuilder: (context, index) => Divider(
          color: Colors.black12,
        ),
        itemCount: playersSorted.length,
        itemBuilder: (context, index) {
          return playerView(index, playersSorted[index]);
        },
      ),
    );
  }

  void _addPlayerLayout() {
    showDialog(
      context: context,
      builder: (BuildContext context) {
        return AlertDialog(
          content: TextField(
            autofocus: true,
            onSubmitted: (val) {
              setState(() {
                players.add(Player(val));
              });
              Navigator.pop(context);
            },
            decoration: new InputDecoration(
              contentPadding: const EdgeInsets.all(16.0),
            ),
          ),
        );
      },
    );
  }

  Widget _playerView(int index, Player player) {
    return Container(
      child: ListTile(
        contentPadding: EdgeInsets.symmetric(vertical: 5.0, horizontal: 16.0),
        title: Text('${players[index].name}'),
        trailing: Wrap(
          spacing: 12,
          children: <Widget>[
            IconButton(
              icon: new Icon(Icons.close),
              onPressed: () {
                setState(() => players.removeAt(index));
              },
            ),
          ],
        ),
      ),
    );
  }

  Widget _playerTotalView(int i, Player p) {
    var score = p.totalScore(this.exLeaders, this.exCities, this.exArmada);
    return new Container(
      child: Row(
        children: [
          Expanded(
            child: ListTile(
              contentPadding:
                  EdgeInsets.symmetric(vertical: 5.0, horizontal: 16.0),
              title: new Text('${p.name}'),
            ),
            flex: 3,
          ),
          Expanded(
            child: ListTile(
              title: new Text(
                '$score',
                textAlign: TextAlign.center,
              ),
            ),
            flex: 1,
          ),
        ],
      ),
    );
  }

  Widget _playerWarView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.warScore,
        min: -50,
        onChange: (v) {
          setState(() {
            p.warScore = v!;
          });
        },
      ),
    );
  }

  Widget _playerMoneyView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.money,
        onChange: (v) {
          setState(() {
            p.money = v!;
          });
        },
      ),
    );
  }

  Widget _playerWonderView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.wonderScore,
        onChange: (v) {
          setState(() {
            p.wonderScore = v!;
          });
        },
      ),
    );
  }

  Widget _playerCivilianView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.civilianScore,
        onChange: (v) {
          setState(() {
            p.civilianScore = v!;
          });
        },
      ),
    );
  }

  Widget _playerCommerceView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.commerceScore,
        onChange: (v) {
          setState(() {
            p.commerceScore = v!;
          });
        },
      ),
    );
  }

  Widget _playerScienceView(int i, Player p) {
    return new Container(
      child: Row(
        children: [
          Expanded(
            flex: 2,
            child: ListTile(
              contentPadding:
                  EdgeInsets.symmetric(vertical: 5.0, horizontal: 16.0),
              title: new Text('${p.name}'),
            ),
          ),
          Expanded(
            child: ListTile(
              title: Container(
                child: TextField(
                  cursorWidth: 0,
                  textAlign: TextAlign.center,
                  keyboardType: TextInputType.number,
                  inputFormatters: [FilteringTextInputFormatter.digitsOnly],
                  controller: TextEditingController()..text = '${p.compass}',
                  onChanged: (v) {
                    setState(() {
                      p.compass = int.tryParse(v.length > 0
                              ? v.replaceFirst("${p.compass}", "")
                              : "0") ??
                          p.compass;
                    });
                  },
                ),
              ),
            ),
          ),
          Expanded(
            child: ListTile(
              title: Container(
                child: TextField(
                  cursorWidth: 0,
                  textAlign: TextAlign.center,
                  keyboardType: TextInputType.number,
                  inputFormatters: [FilteringTextInputFormatter.digitsOnly],
                  controller: TextEditingController()..text = '${p.gears}',
                  onChanged: (v) {
                    setState(() {
                      p.gears = int.tryParse(v.length > 0
                              ? v.replaceFirst("${p.gears}", "")
                              : "0") ??
                          p.gears;
                    });
                  },
                ),
              ),
            ),
          ),
          Expanded(
            child: ListTile(
              title: Container(
                child: TextField(
                  cursorWidth: 0,
                  textAlign: TextAlign.center,
                  keyboardType: TextInputType.number,
                  inputFormatters: [FilteringTextInputFormatter.digitsOnly],
                  controller: TextEditingController()..text = '${p.tablets}',
                  onChanged: (v) {
                    setState(() {
                      p.tablets = int.tryParse(v.length > 0
                              ? v.replaceFirst("${p.tablets}", "")
                              : "0") ??
                          p.tablets;
                    });
                  },
                ),
              ),
            ),
          ),
          Expanded(
            child: ListTile(
              title: Container(
                child: TextField(
                  cursorWidth: 0,
                  textAlign: TextAlign.center,
                  keyboardType: TextInputType.number,
                  inputFormatters: [FilteringTextInputFormatter.digitsOnly],
                  controller: TextEditingController()..text = '${p.wilds}',
                  onChanged: (v) {
                    setState(() {
                      p.wilds = int.tryParse(v.length > 0
                              ? v.replaceFirst("${p.wilds}", "")
                              : "0") ??
                          p.wilds;
                    });
                  },
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }

  Widget _playerGuildeView(int i, Player p) {
    int max = 999;
    if (this.exArmada) {
      max = 10;
      p.guildeScore = min(p.guildeScore, 10);
    }
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.guildeScore,
        max: max,
        onChange: (v) {
          setState(() {
            p.guildeScore = v!;
          });
        },
      ),
    );
  }

  Widget _playerLeadersView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.leaders,
        onChange: (v) {
          setState(() {
            p.leaders = v!;
          });
        },
      ),
    );
  }

  Widget _playerCitiesView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.cities,
        onChange: (v) {
          setState(() {
            p.cities = v!;
          });
        },
      ),
    );
  }

  Widget _playerDebtView(int i, Player p) {
    if (p.debt > 0) p.debt = 0;
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.debt,
        max: 0,
        min: -100,
        onChange: (v) {
          setState(() {
            p.debt = v!;
          });
        },
      ),
    );
  }

  Widget _playerArmadaView(int i, Player p) {
    return new ListTilePlayerWithNumberSelector(
      playerName: p.name!,
      selector: NumberSelector(
        value: p.armada,
        onChange: (v) {
          setState(() {
            p.armada = v!;
          });
        },
      ),
    );
  }
}
